app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/gainorlossafterreliefs/GainOrLossAfterReliefsController.scala

```diff
@@ -62,21 +66,52 @@ class GainOrLossAfterReliefsController @Inject() (
 
   import GainOrLossAfterReliefsController._
 
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
   def enterGainOrLossAfterReliefs(): Action[AnyContent] =
     authenticatedActionWithSessionData.async { implicit request =>
       withFillingOutReturnAndAnswers { (fillingOutReturn, draftReturn, answer) =>
-        Ok(
-          gainOrLossAfterReliefsPage(
-            answer.fold(gainOrLossAfterReliefsForm)(value => gainOrLossAfterReliefsForm.fill(value.inPounds())),
-            answer.fold(controllers.returns.routes.TaskListController.taskList())(_ =>
-              routes.GainOrLossAfterReliefsController.checkYourAnswers()
-            ),
-            fillingOutReturn.subscribedDetails.isATrust,
-            draftReturn.representativeType(),
-            draftReturn.triageAnswers().isLeft,
-            fillingOutReturn.isAmendReturn
+        glarCalculatorEligibilityUtil
+          .isEligibleForFurtherReturnOrAmendCalculation(fillingOutReturn)
+          .fold(
+            err => {
+              logger.warn("Could not check for calculation eligibility", err)
+              errorHandler.errorResult()
+            },
+            furtherReturnEligibility => {
+              furtherReturnEligibility.previousReturnsImplyEligibility match {
+                case Some(previousReturnsImplyEligibility)
+                    if !fillingOutReturn.previousReturnsImplyEligilityForFurtherReturnCalculation
+                      .contains(previousReturnsImplyEligibility) =>
+                  val updatedJourney = fillingOutReturn.copy(previousReturnsImplyEligilityForFurtherReturnCalculation =
+                    Some(previousReturnsImplyEligibility)
+                  )
+                  for {
+                    _ <- returnsService.storeDraftReturn(updatedJourney)
+                    _ <- EitherT(
+                           updateSession(sessionStore, request)(
+                             _.copy(journeyStatus = Some(updatedJourney))
+                           )
+                         )
+                  } yield ()
+                case _ => ()
+              }
+              Ok(
+                gainOrLossAfterReliefsPage(
+                  answer.fold(gainOrLossAfterReliefsForm)(value => gainOrLossAfterReliefsForm.fill(value.inPounds())),
+                  answer.fold(controllers.returns.routes.TaskListController.taskList())(_ =>
+                    routes.GainOrLossAfterReliefsController.checkYourAnswers()
+                  ),
+                  fillingOutReturn.subscribedDetails.isATrust,
+                  draftReturn.representativeType(),
+                  draftReturn.triageAnswers().isLeft,
+                  fillingOutReturn.isAmendReturn,
+                  if (furtherReturnEligibility.isEligible)
```

Comment:
`furtherReturnEligibility.isEligible` and `None` shouldn't be an allowable state here. One way to avoid it is to maybe have something like:
```scala
sealed trait FurtherReturnEligibility

object FurtherReturnEligibility {
  case class Eligible(calculation: CalculatedGlarBreakdown) // previousReturnsImplyEligility must be true here
  case class Ineligible(previousReturnsImplyEligility: Option[Boolean])
}
```
and move the calculation logic to the service